home/tests/test_serializers.py

```diff
@@ -0,0 +1,388 @@
+from typing import List, Union
+
+import pytest
+
+from home.models import ContentPage, HomePage
+from home.serializers import (
+    format_whatsapp_message,
+    has_next_message,
+    has_previous_message,
+)
+
+from .page_builder import (
+    MBlk,
+    MBody,
+    PageBuilder,
+    SBlk,
+    SBody,
+    UBlk,
+    UBody,
+    VarMsg,
+    VBlk,
+    VBody,
+    WABlk,
+    WABody,
+)
+
+
+@pytest.mark.django_db
+class TestHasNextMessage:
+    def create_content_page(
+        self,
+        parent: Union[ContentPage, None] = None,
+        title: str = "default page",
+        tags: Union[List[str], None] = None,
+        wa_body_count: int = 1,
+        publish: bool = True,
+    ) -> ContentPage:
+        """
+        Helper function to create pages needed for each test.
+
+        Parameters
+        ----------
+        parent : ContentPage
+            The ContentPage that will be used as the parent of the content page.
+
+            If this is not provided, a ContentPageIndex object is created as a child of
+            the default home page and that is used as the parent.
+        title : str
+            Title of the content page.
+        tags : [str]
+            List of tags on the content page.
+        wa_body_count : int
+            How many WhatsApp message bodies to create on the content page.
+        publish: bool
+            Should the content page be published or not.
+        """
+        if not parent:
+            home_page = HomePage.objects.first()
+            main_menu = PageBuilder.build_cpi(home_page, "main-menu", "Main Menu")
+            parent = main_menu
+
+        bodies = [
+            MBody(title, [MBlk("*Default Messenger Content* 🏥")]),
+            SBody(title, [SBlk("*Default SMS Content* ")]),
+            UBody(title, [UBlk("*Default USSD Content* ")]),
+            VBody(title, [VBlk("*Default Viber Content* ")]),
+        ]
+
+        for i in range(wa_body_count):
+            wa_body = f"*Default WhatsApp Content {i+1}* 🏥"
+            bodies.append(WABody(title, [WABlk(wa_body)]))
+
+        content_page = PageBuilder.build_cp(
+            parent=parent,
+            slug=title.replace(" ", "-"),
+            title=title,
+            bodies=bodies,
+            tags=tags or [],
+            quick_replies=[],
+            triggers=[],
+            publish=publish,
+        )
+        return content_page
+
+    # Test has_next_message
+    def test_has_next_message_whatsapp(self) -> None:
+        """Checks if the next message matches up.
+        Index of first message is 0, next message is then message 2 with index 1"""
+        page = self.create_content_page(wa_body_count=2)
+        assert page.whatsapp_body
+        assert has_next_message(0, page, "whatsapp") == 2
+
+    def test_has_next_message_viber_no_next_message(self) -> None:
+        """On a page with 2 viber messages, the next message is None"""
```

Comment:
Do we have two viber messages here? Also, is the `None` because we're looking at viber messages, or because we pass index 1 to `has_next_message`?

Reply:
Oops, "on a page with 1 viber message"

Reply:
This docstring is still confusing, as is the one above which is identical despite having a different message index.

As I understand it, `has_next_message()` should return the index of the message after the given index (for the given platform) if there is one, otherwise `None`. If so, perhaps these might be better docstrings given the code for these two tests:

"The next message after the last message is None."
"The next message after a message that doesn't exist is None."

However, one of the test names indicates that we should get a non-None result? 🤷

Reply:
damn I just realised that I forgot that the message number != index, thursday meg seems to have been very very unfocused, apologies, fixing these two tests

Reply:
This is why mixing zero-indexed and one-indexed values is such a terrible idea. It basically guarantees this sort of mistake. 🤷